titus-server-master/src/main/java/com/netflix/titus/master/kubernetes/controller/NodeGcController.java

```diff
@@ -0,0 +1,180 @@
+                            AgentManagementService agentManagementService,
+                            KubeApiFacade kubeApiFacade,
+                            LocalScheduler scheduler) {
+        this.configuration = configuration;
+        this.agentManagementService = agentManagementService;
+        this.kubeApiFacade = kubeApiFacade;
+        this.scheduler = scheduler;
+
+        this.clock = titusRuntime.getClock();
+        this.nodesToGc = titusRuntime.getRegistry().gauge(METRIC_ROOT + "nodesToGc");
+    }
+
+    @Activator
+    public void enterActiveMode() {
+        ScheduleDescriptor reconcileSchedulerDescriptor = ScheduleDescriptor.newBuilder()
+                .withName("nodeGcController")
+                .withDescription("GC nodes that no longer exist")
+                .withInitialDelay(Duration.ofMillis(configuration.getNodeGcControllerInitialDelayMs()))
+                .withInterval(Duration.ofMillis(configuration.getNodeGcControllerIntervalMs()))
+                .withTimeout(Duration.ofMillis(configuration.getNodeGcControllerTimeoutMs()))
+                .build();
+        scheduler.schedule(reconcileSchedulerDescriptor, e -> gcNodes(), ExecutorsExt.namedSingleThreadExecutor("node-gc-controller"));
+    }
+
+    void gcNodes() {
+        if (!kubeApiFacade.getNodeInformer().hasSynced()) {
+            logger.info("Node informer has not synced");
+            return;
+        }
+
+        if (!configuration.isNodeGcControllerEnabled()) {
+            logger.info("Node GC is disabled");
+            return;
+        }
+
+        List<V1Node> nodes = kubeApiFacade.getNodeInformer().getIndexer().list();
+        Set<String> nodeNamesToGc = nodes.stream()
+                .filter(this::isNodeEligibleForGc)
+                .map(n -> KubeUtil.getMetadataName(n.getMetadata()))
+                .filter(StringExt::isNotEmpty)
+                .collect(Collectors.toSet());
+
+        logger.info("Attempting to GC {} nodes: {}", nodeNamesToGc.size(), nodeNamesToGc);
+        this.nodesToGc.set(nodeNamesToGc.size());
+        for (String nodeName : nodeNamesToGc) {
+            gcNode(nodeName);
+        }
+        logger.info("Finished node GC");
+    }
+
+    @VisibleForTesting
+    boolean isNodeEligibleForGc(V1Node node) {
+        String nodeName = KubeUtil.getMetadataName(node.getMetadata());
+        Optional<V1NodeCondition> readyNodeConditionOpt = KubeUtil.findNodeCondition(node, READY);
+        if (StringExt.isNotEmpty(nodeName) && readyNodeConditionOpt.isPresent()) {
+            V1NodeCondition readyNodeCondition = readyNodeConditionOpt.get();
+            boolean isReadyConditionTimestampPastGracePeriod = hasConditionGracePeriodElapsed(readyNodeCondition, configuration.getNodeGcGracePeriodMs());
+            boolean isAgentInstanceNotAvailable = isAgentInstanceNotAvailable(nodeName);
+            return isReadyConditionTimestampPastGracePeriod && isAgentInstanceNotAvailable;
```

Comment:
very minor optimization here: inline the second variable and prevent some wasted cycles when the first check already fails. The method call is already very descriptive and readability isn't much improved by the extra variable.